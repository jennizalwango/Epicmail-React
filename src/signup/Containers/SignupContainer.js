import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Signup from '../Components/SignupComponent';
import signupUtil from '../redux/validations';
import signupAction from '../redux/actions/SignupAction';


export class SignupContainer extends Component {
    constructor (props) {
        super(props);
        this.state = {
            firstname: '',
            lastname: '',
            email: '',
            password: '',
            errors: {
                invalid_email: false
            }
        };
    }

  onChange = (event) => {
      this.setState({ [event.target.name]: event.target.value });

      const { target } = event;
      const { name, value } = target;
      const { errors } = this.state;
      switch (name) {
          case 'email':
              if (!this.checkValidEmail(value)) {
                  this.setState({ errors: { userEmail: 'Please provide a valid email address' } });
                  errors.invalid_email = true;
              } else {
                  this.setState({ errors: { userEmail: '' } });
                  errors.invalid_email = false;
              }
              break;
          default:
              break;
      }

      this.setState({ [event.target.name]: event.target.value });
  }

  checkValidEmail = (userEmail) => {
      const mailFormat = signupUtil.SIGNUP_UTILS;
      if (userEmail.match(mailFormat)) {
          return true;
      }
      return false;
  };

  handleSubmit = (event) => {
      event.preventDefault();
      const { email, password, firstname, lastname } = this.state;
      const signupdata = {
          first_name: firstname,
          last_name: lastname,
          email,
          password,
          isAdmin: true
      };

      if (email === '') {
          this.setState({ errors: { userEmail: 'Email field required' } });
      }
      if (password === '') {
          this.setState({ errors: { userPassword: 'Password field is required' } });
      }
      if (firstname === '') {
          this.setState({ errors: { userfirstname: 'firstname field required' } });
      }
      if (lastname === '') {
          this.setState({ errors: { userlastname: 'lastname field is required' } });
      }
      const { propsSignupAction } = this.props;

      const { errors } = this.state;
      if (!errors.invalid_email && errors.userEmail === '') {
          propsSignupAction(signupdata, this.props);
      }
  }

  render () {
      const { email, password, firstname, lastname, errors } = this.state;
      const {
          SignupFailed,
          isLoading,
          isSigningup
      } = this.props;
      return (
          <Signup
              handleSubmit={this.handleSubmit}
              onChange={this.onChange}
              email={email}
              password={password}
              firstname={firstname}
              lastname={lastname}
              errors={errors}
              isSigningup={isSigningup}
              SignupFailed={SignupFailed}
              isLoading={isLoading}
          />
      );
  }
}

SignupContainer.propTypes = {
    propsSignupAction: PropTypes.func.isRequired,
    isLoading: PropTypes.bool.isRequired,
    isSigningup: PropTypes.bool.isRequired,
    SignupFailed: PropTypes.bool.isRequired
};

const mapStateToProps = state => ({
    signup: state.signup,
    isSigningup: state.signup.is_signing_up,
    iSignupFailed: state.signup.signup_failed,
    isLoading: state.signup.isLoading
});
export default connect(mapStateToProps, { propsSignupAction: signupAction })(SignupContainer);
