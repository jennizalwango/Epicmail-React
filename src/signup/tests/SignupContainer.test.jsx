import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow, mount } from 'enzyme';
import { SignupContainer } from '../Containers/SignupContainer';

describe('SignupContainer Component', () => {
    const props = {
        propsSignupAction: jest.fn(),
        isLoading: false,
        isSigningup: false,
        SignupFailed: false
    };

    const event = {
        preventDefault: () => {},
        target: {
            name: 'email',
            value: 'abc'
        }
    };
    const wrapper = shallow(<SignupContainer {...props} />);

    it('matches snapshot', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should call an onChange function', () => {
        const component = shallow(
            <SignupContainer {...props} />,
        );
        component.setState({ email: 'email' });
        component.instance().onChange(event);
        expect(component.instance().state.email).toBe('abc');
    });

    it('should call handleSubmit', () => {
        const component = mount(
            <SignupContainer {...props} />,
        );
        const event = {
            preventDefault: () => {},
        };
        const instance = component.instance();
        const handleSubmit = jest.spyOn(instance, 'handleSubmit');
        instance.handleSubmit({ preventDefault: jest.fn() });
        expect(handleSubmit).toHaveBeenCalled();
    });

    it('should call a submit a form', () => {
        wrapper.setState({
            email: 'jennyj'
        });
        const userEmail = {
            match: jest.fn()
        };
        const instance = wrapper.instance();
        const checkValidEmail = jest.spyOn(instance, 'checkValidEmail');
        checkValidEmail(userEmail);
        instance.handleSubmit({ preventDefault: jest.fn() });
        expect(checkValidEmail).toHaveBeenCalled();
    });

    it('Valid Email', () => {
        wrapper.setState({
            email: 'jenny@gmail.com'
        });
        const userEmail = {
            match: jest.fn()
        };
        const instance = wrapper.instance();
        const checkValidEmail = jest.spyOn(instance, 'checkValidEmail');
        instance.handleSubmit({ preventDefault: jest.fn() });
        checkValidEmail(userEmail);
        expect(checkValidEmail).toHaveBeenCalled();
    });

    it('should call submit form', () => {
        wrapper.setState({
            firstname: 'jennyj',
            lastname: 'zalwango',
            email: 'jennyj',
            password: 'password'
        });
        const userEmail = {
            match: jest.fn()
        };
        const instance = wrapper.instance();
        const checkValidEmail = jest.spyOn(instance, 'checkValidEmail');
        instance.handleSubmit({ preventDefault: jest.fn() });
        checkValidEmail(userEmail);
        expect(checkValidEmail).toHaveBeenCalled();
    });

    it('should validate an empty password', () => {
        
        const instance = wrapper.instance();
        instance.setState({
            firstname: 'jennyj',
            lastname: 'zalwango',
            email: 'jenny@gmail.com',
            password: ''
        });
        instance.handleSubmit({ preventDefault: jest.fn() });
        expect('Password field is required').toEqual(instance.state.errors.userPassword);
    });

    it('should call loginAction when handling submit', () => {
        wrapper.setState({
            email: 'haki@gmail.com',
            password: 'password',
            errors: {}
        });
        const instance = wrapper.instance();
        instance.handleSubmit({ preventDefault: jest.fn() });
        props.propsSignupAction();
        expect(props.propsSignupAction).toHaveBeenCalled();
    });

    it('should load the component', () => {
        expect(wrapper).toHaveLength(1);
    });
});
