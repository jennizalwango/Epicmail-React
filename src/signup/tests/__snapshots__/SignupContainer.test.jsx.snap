// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupContainer Component matches snapshot 1`] = `
<Signup
  SignupFailed={false}
  email=""
  errors={
    Object {
      "invalid_email": false,
    }
  }
  firstname=""
  handleSubmit={[Function]}
  isLoading={false}
  isSigningup={false}
  lastname=""
  onChange={[Function]}
  password=""
/>
`;
